Bugzilla/App/BMO/ComponentGraveyard.pm

```diff
@@ -0,0 +1,311 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::App::BMO::ComponentGraveyard;
+use Mojo::Base 'Mojolicious::Controller';
+
+use Bugzilla;
+use Bugzilla::Component;
+use Bugzilla::Constants;
+use Bugzilla::Error;
+use Bugzilla::Field;
+use Bugzilla::Hook;
+use Bugzilla::Logging;
+use Bugzilla::Product;
+use Bugzilla::Token;
+use Bugzilla::Util;
+
+sub setup_routes {
+  my ($class, $r) = @_;
+  $r->any('/admin/component/graveyard')->to('BMO::ComponentGraveyard#graveyard')
+    ->name('component_graveyard');
+}
+
+sub graveyard {
+  my ($self) = @_;
+  Bugzilla->usage_mode(USAGE_MODE_MOJO);
+  my $user = $self->bugzilla->login(LOGIN_REQUIRED) || return undef;
+
+  my $editcomponents    = $user->in_group('editcomponents');
+  my $editable_products = $user->get_products_by_permission('editcomponents');
+
+  $editcomponents
+    || scalar @{$editable_products}
+    || return $self->user_error('auth_failure',
+    {group => 'editcomponents', action => 'edit', object => 'components'});
+
+  # # Discover all provided parameters such as component and product
+  my $product_name   = $self->param('product')   || '';
+  my $component_name = $self->param('component') || '';
+  my $token          = $self->param('token');
+
+  if ($self->req->method eq 'POST') {
+    my $post_params = $self->req->body_params->to_hash;
+    $product_name   = $post_params->{product}   || '';
+    $component_name = $post_params->{component} || '';
+    $token          = $post_params->{token};
+  }
+
+  # If the user has editcomponents privs, then we include
+  # all selectable products
+  if ($editcomponents) {
+    $editable_products = $user->get_selectable_products;
+  }
+
+  # Remove current graveyard products from product list we do
+  # not want to move components out of the graveyard
+  $editable_products = [grep { $_->name !~ /Graveyard$/ } @{$editable_products}];
+
+  my $vars = {editable_products => $editable_products,};
+  $vars->{selected_product}   = $product_name   if $product_name;
+  $vars->{selected_component} = $component_name if $component_name;
+
+  # Redisplay the product/component selection form if a product
+  # and component were not selected
+  if (!$product_name || !$component_name) {
+    $vars->{token} = issue_session_token('component_graveyard');
+    $self->stash(%{$vars});
+    return $self->render(
+      template => 'admin/components/graveyard',
+      handler  => 'bugzilla'
+    );
+  }
+
+  # Handle CSRF tokens
+  check_token_data($token, 'component_graveyard');
+  delete_token($token);
+
+  # Sanity checks first
```

Comment:
The sanity checks also load the $product and $component objects needed later which would need to be returned from the function. I left it in since we moved a large part of the later code to Bugzilla::Component so it is already significantly shorter.